lib/Conversion/BGVToPoly/BGVToPoly.cpp

```diff
@@ -0,0 +1,83 @@
+#include "include/Conversion/BGVToPoly/BGVToPoly.h"
+
+#include <cstddef>
+#include <optional>
+
+#include "include/Dialect/BGV/IR/BGVDialect.h"
+#include "include/Dialect/BGV/IR/BGVTypes.h"
+#include "include/Dialect/Poly/IR/PolyAttributes.h"
+#include "include/Dialect/Poly/IR/PolyTypes.h"
+#include "include/Dialect/Poly/IR/Polynomial.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"                  // from @llvm-project
+#include "mlir/Dialect/Func/Transforms/FuncConversions.h"  // from @llvm-project
+#include "mlir/Dialect/Func/Transforms/OneToNFuncConversions.h"  // from @llvm-project
+#include "mlir/Dialect/Tensor/IR/Tensor.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/DialectConversion.h"  // from @llvm-project
+
+namespace mlir::heir::bgv {
+
+#define GEN_PASS_DEF_BGVTOPOLY
+#include "include/Conversion/BGVToPoly/BGVToPoly.h.inc"
+
+class CiphertextTypeConverter : public TypeConverter {
+ public:
+  // Convert ciphertext to tensor<#dim x !poly.poly<#rings[#level]>>
+  CiphertextTypeConverter(MLIRContext *ctx) {
+    addConversion([](Type type) { return type; });
+    addConversion([ctx](CiphertextType type) -> Type {
+      assert(type.getLevel().has_value());
+      auto level = type.getLevel().value();
+      assert(level < type.getRings().getRings().size());
+
+      auto ring = type.getRings().getRings()[level];
+      auto polyTy = poly::PolynomialType::get(ctx, ring);
+
+      return RankedTensorType::get({type.getDim()}, polyTy);
+    });
+  }
+  // We don't include any custom materialization ops because this lowering is
+  // all done in a single pass. The dialect conversion framework works by
+  // resolving intermediate (mid-pass) type conflicts by inserting
+  // unrealized_conversion_cast ops, and only converting those to custom
+  // materializations if they persist at the end of the pass. In our case,
+  // we'd only need to use custom materializations if we split this lowering
+  // across multiple passes.
+};
+
+struct BGVToPoly : public impl::BGVToPolyBase<BGVToPoly> {
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    auto *module = getOperation();
+    CiphertextTypeConverter typeConverter(context);
+
+    ConversionTarget target(*context);
+    target.addLegalOp<ModuleOp>();
+
+    RewritePatternSet patterns(context);
+
+    // Add "standard" set of conversions and constraints for full dialect
+    // conversion. See Dialect/Func/Transforms/FuncBufferize.cpp for example.
+    populateFunctionOpInterfaceTypeConversionPattern<func::FuncOp>(
+        patterns, typeConverter);
+    target.addDynamicallyLegalOp<func::FuncOp>([&](func::FuncOp op) {
+      return typeConverter.isSignatureLegal(op.getFunctionType()) &&
+             typeConverter.isLegal(&op.getBody());
+    });
+
+    populateBranchOpInterfaceTypeConversionPattern(patterns, typeConverter);
```

Comment:
if it's easy, we may want to add `populateCallOpTypeConversionPattern` (I expect we'll probably at some point have submodules to optimize) - we can leave this in for later though, it'll be obvious when we use multiple functions in a module that it needs to be added.